From 599bc8c0e9965f4e300fd3cba7038fe7893768d9 Mon Sep 17 00:00:00 2001
From: Hadar Hen Zion <hadarh@mellanox.com>
Date: Thu, 21 Mar 2013 05:55:53 +0000
Subject: [PATCH 534/858] net/mlx4_en: Remove ethtool flow steering rules
 before releasing QPs

Fix the ethtool flow steering rules cleanup to be carried out before
releasing the RX QPs.

Signed-off-by: Hadar Hen Zion <hadarh@mellanox.com>
Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 6efb5fac4d6b617972ab5a10bf67e0eba2c2d212)

Signed-off-by: Arnaldo Carvalho de Melo <acme@redhat.com>
---
 drivers/net/ethernet/mellanox/mlx4/en_netdev.c | 22 +++++++++++-----------
 1 file changed, 11 insertions(+), 11 deletions(-)

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_netdev.c b/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
index 5c4b61a..8723b8e 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
@@ -1617,6 +1617,17 @@ void mlx4_en_stop_port(struct net_device *dev, int detach)
  /* Flush multicast filter */
  mlx4_SET_MCAST_FLTR(mdev->dev, priv->port, 0, 1, MLX4_MCAST_CONFIG);
 
+ /* Remove flow steering rules for the port*/
+ if (mdev->dev->caps.steering_mode ==
+     MLX4_STEERING_MODE_DEVICE_MANAGED) {
+  ASSERT_RTNL();
+  list_for_each_entry_safe(flow, tmp_flow,
+      &priv->ethtool_list, list) {
+   mlx4_flow_detach(mdev->dev, flow->id);
+   list_del(&flow->list);
+  }
+ }
+
  mlx4_en_destroy_drop_qp(priv);
 
  /* Free TX Rings */
@@ -1637,17 +1648,6 @@ void mlx4_en_stop_port(struct net_device *dev, int detach)
  if (!(mdev->dev->caps.flags2 & MLX4_DEV_CAP_FLAGS2_REASSIGN_MAC_EN))
   mdev->mac_removed[priv->port] = 1;
 
- /* Remove flow steering rules for the port*/
- if (mdev->dev->caps.steering_mode ==
-     MLX4_STEERING_MODE_DEVICE_MANAGED) {
-  ASSERT_RTNL();
-  list_for_each_entry_safe(flow, tmp_flow,
-      &priv->ethtool_list, list) {
-   mlx4_flow_detach(mdev->dev, flow->id);
-   list_del(&flow->list);
-  }
- }
-
  /* Free RX Rings */
  for (i = 0; i < priv->rx_ring_num; i++) {
   mlx4_en_deactivate_rx_ring(priv, &priv->rx_ring[i]);
-- 
1.8.3.1